src/hotspot/share/services/virtualMemoryTracker.cpp

```diff
@@ -523,29 +525,72 @@ bool VirtualMemoryTracker::remove_released_region(address addr, size_t size) {
     }
   }
 
-  VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
-
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  if (reserved_rgn->base() == addr ||
-      reserved_rgn->end() == addr + size) {
+  if (size <= reserved_rgn->size() && reserved_rgn->contain_region(addr, size)) {
+    // The requested chunk to be freed is smaller or same size as the real
+    // underlaying region and is fully contained within one (i.e. single region)
+    VirtualMemorySummary::record_released_memory(size, reserved_rgn->flag());
+    assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
+    if (reserved_rgn->base() == addr || reserved_rgn->end() == addr + size) {
```

Comment:
This is the one I wanted a comment for! "If the two regions start at the same place, or ends at the same place, then we can shrink the region from the left or right. Otherwise, we'll have to split it up"

Reply:
We still have to split in some cases where the two regions start at the same place, i.e.:

```
aaaa bbbb cccc
xxxx xxxx
```

Where a,b,c are the underlaying strips, and x denotes the chunk we want to free.